packages/core/admin/ee/server/services/review-workflows/stages.js

```diff
@@ -1,36 +1,88 @@
 'use strict';
 
+const { mapAsync } = require('@strapi/utils');
+
 const { STAGE_MODEL_UID } = require('../../constants/workflows');
+const { getService } = require('../../utils');
+
+module.exports = ({ strapi }) => {
+  const workflowsService = getService('workflows', { strapi });
+
+  return {
+    find({ workflowId, populate }) {
+      const params = {
+        filters: { workflow: workflowId },
+        populate,
+      };
+      return strapi.entityService.findMany(STAGE_MODEL_UID, params);
+    },
+
+    findById(id, { workflowId, populate }) {
+      const params = {
+        filters: { workflow: workflowId },
+        populate,
+      };
+      return strapi.entityService.findOne(STAGE_MODEL_UID, id, params);
+    },
+
+    createMany(stagesList, { fields }) {
+      const params = {
+        select: fields,
+      };
+      return Promise.all(
+        stagesList.map((stage) =>
+          strapi.entityService.create(STAGE_MODEL_UID, { data: stage, ...params })
+        )
+      );
+    },
+
+    update(stageId, stageData) {
+      return strapi.entityService.update(STAGE_MODEL_UID, stageId, { data: stageData });
+    },
+
+    delete(stageId) {
+      return strapi.entityService.delete(STAGE_MODEL_UID, stageId);
+    },
+
+    count() {
+      return strapi.entityService.count(STAGE_MODEL_UID);
+    },
+
+    async replaceWorkflowStages(workflowId, stages) {
+      const workflow = await workflowsService.findById(workflowId, { populate: ['stages'] });
+
+      const { created, updated, deleted } = getDiffBetweenStages(workflow.stages, stages);
+
+      const newStages = await this.createMany(created, { fields: ['id'] });
+      const assignedStagesToWorkflow = stages.map((stage) => stage.id ?? newStages.shift().id);
+
+      await mapAsync(updated, (stage) => this.update(stage.id, stage));
+      await mapAsync(deleted, (stage) => this.delete(stage.id));
+      return workflowsService.update(workflowId, {
+        stages: assignedStagesToWorkflow,
+      });
+    },
+  };
+};
+
+function getDiffBetweenStages(sourceStages, comparisonStages) {
```

Comment:
I think this might benefit from some docs. Something like:

```
/**
 * Determine the stages that need creating, updating or deleting by comparing
 * the current stage state with the new stages received
 * @typedef {{id: Number, name: String, workflow: Number}} Stage
 * @type {(
 * sourceStages: Array.<Stage>,
 * comparisonStages: Array.<Stage>
 * ) => {
 * created: Array.<Stage>,
 * updated: Array.<Stage>,
 * deleted: Array.<Stage>,
 * }}
 */
```
wdyt?

Reply:
Sure

Reply:
Very nice, I think that comment helps a lot